### PR TITLE
Throttle SetEvent call in render thread

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

This is an attempt to reduce unnecessary `SetEvent` CPU overhead.

 
## PR Checklist
* [ ] Closes #xxx
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

Redrawing is triggered intensively when terminal is under load. The code here indicates that the redraw call will be throttled:

https://github.com/microsoft/terminal/blob/94fc40ed31c0de050304248644458b21d0cce2e9/src/renderer/base/renderer.cpp#L154

Still, the `SetEvent` call itself stands out a 7% usage of CPU in Release build. Those `SetEvent` calls are mostly useless because most of them are invoked while the rendering thread is sleeping (for `s_FrameLimitMilliseconds` long). 

图片

I've tried to guard all the useless calls with `_fPainting` flag, as in this PR. It works great for the CPU. But the rendering became noticeably sluggish. 

I've figured that the rendering thread design is flawed in the perspective of this certain issue, probably also future unknown issues. Instead of being a thread that can be notified, it should behave more like a *timer* that invokes rendering opereation every `s_FrameLimitMilliseconds`. In this way, the rendering thread will be more focused on its own work without being distracted from others.

Maybe @egmontkob can provide us some insight about this.

### CODE REVIEW

File: src/renderer/base/thread.hpp [hunk 1]
```diff
@@ -47,5 +47,6 @@ namespace Microsoft::Console::Render
         IRenderer* _pRenderer; // Non-ownership pointer
 
         bool _fKeepRunning;
+        volatile bool _fPainting;
```
Comment (zadjii-msft) [inline src/renderer/base/thread.hpp]: should we use a `std::atomic` here instead of a `volatile`?

**Comment by zadjii-msft** [comment]
Maybe the sluggishness comes from us ignoring all frame triggers during the paint. Previously, if something were to trigger a new frame _during_ a paint, then we'd make sure to start another frame right away on the next timer tick. Now, with just the right timing, if there's not any action between one frame and the next timer tick, then that frame is now skipped until something fully _after_ the first frame triggers a new frame. 

Maybe if we kept another flag, which is set when `TriggerRedraw` is called _during_ a frame, we could immediately call `SetEvent` again at the end of a frame, if something else requested a new frame _during_ that frame.

**Comment by skyline75489** [comment]
@zadjii-msft Thanks for the idea. The sluggishness is gone indeed. I think it's good enough to out of draft now.

**Comment by skyline75489** [comment]
According to cppreference https://en.cppreference.com/w/cpp/atomic/memory_order, quote

> One notable exception is Visual Studio, where, with default settings, every volatile write has release semantics and every volatile read has acquire semantics 

Seems volatile is also good for this.

**Comment by DHowett-MSFT** [comment]
Voicing disagreement with using `volatile`. If we want specific memory orders, we should _request_ them. `volatile` is the subject of an active [proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1152r0.html) for deprecation, and I don’t think net new code should be taking a dependency on it.

File: src/renderer/base/thread.cpp [hunk 2]
```diff
@@ -158,26 +160,46 @@ DWORD WINAPI RenderThread::_ThreadProc()
     while (_fKeepRunning)
     {
         WaitForSingleObject(_hPaintEnabledEvent, INFINITE);
-        WaitForSingleObject(_hEvent, INFINITE);
+
+        // Skip waiting if next frame is requested.
+        if (_fNextFrameRequested.test_and_set(std::memory_order_relaxed))
```
Comment (skyline75489) [inline src/renderer/base/thread.cpp]: I think this can be relaxed since it happens mostly during when the render thread is sleeping.

**Comment by skyline75489** [comment]
@DHowett-MSFT Thanks for the clarification. I came from a Java background and they also have `volatile` in Java. Just learned that essentially that is different from `volatile` in C/C++.

**Comment by skyline75489** [comment]
图片

@zadjii-msft There you go. Pretty good huh?

**Comment by zadjii-msft** [comment]
@skyline75489 This does put a smile on my face ☺️ Thanks!

**Comment by miniksa** [comment]
@skyline75489, sorry this took me so so so long to get around to. I just started digging back into rendering and performance things this week and am coming back around to these PRs. This looks wonderful to me and I approve. Thanks for your effort and sorry for the delay.

**Comment by DHowett-MSFT** [comment]
:tada: Once again, thanks for the contribution!

This pull request was included in a set of conhost changes that was just
released with [Windows Insider Build 19603](https://blogs.windows.com/windowsexperience/2020/04/08/announcing-windows-10-insider-preview-build-19603/).

**Pull request merged**